docs/rfc/rc-005-event-system.rst

```diff
@@ -0,0 +1,121 @@
+=====================
+RFC 005: Event System
+=====================
+
+Changelog
+---------
+
+- 2021-09-17: Initial Draft (@tychoish)
+
+Abstract
+--------
+
+The event system within Tendermint, which supports a lot of core
+functionality, also represents a major infrastructural liability. As part of
+our upcoming review of the RPC interfaces and our ongoing thoughts about
+stability and performance, as well as the preparation for Tendermint 1.0, we
+should revisit the design and implementation of the event system. This
+document discusses both the current state of the system 
+
+Background
+----------
+
+Current State of Events
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The event system makes it possible for clients, both internal and external,
+to receive notifications after a series of given events, such as new blocks,
+new transactions, validator set changes, as well as intermediate events during
+consensus. Because the event system is very cross cutting, the behavior and
+performance of the event publication and subscription system has huge impacts
+for all of Tendermint.
+
+The subscription service is exposed over the RPC interface, but also powers
+the indexing (e.g. to an external database,) and is the mechanism by which
+`BroadcastTxCommit` is able to wait for transactions to land in a block.
```

Comment:
I think this is all mostly true, but there are a few things I think worth calling out, relevant to how the various components interact with each other:

1. I believe indexing doesn't require _subscribing_ to the event stream. I think events are written to the sinks as the node is generating and publishing the events, i.e., pre-pubsub.
2. Event subscription shares plumbing with the RPC service, but I believe is _only_ exported via the websocket endpoint, not gRPC or the HTTP (URL/JSON). That detail isn't crucial except to note that event subscription may be separable from the rest of the RPC machinery with some work (I'm working on a proposal for maybe doing that).
3. I believe the mempool manages the subscription used to wait for `BroadcastTxCommit` rather than the RPC service (the latter does wind up blocking on the result, but the streaming happens internally on this path).

Reply:
1. Are indexing events buffered/dispatched independently, or could one of these systems put back pressure on the other. 
2. Sweet.
3. https://github.com/tendermint/tendermint/blob/master/rpc/core/mempool.go#L71-L86 it really looks like this happens in the rpc end point.

Reply:
1. It's not clear. The code makes it look like they are done independently, but the way the channels are set up, I think it's possible publishing _could_ block indexing. I'd have to trace it out, but I think the important bit is that it isn't expected to be dependent (except in order).
3. Ah yes, indeed it does.

Reply:
So, I was wrong about my original point (1) above. The indexer _service_ works by subscribing to the event stream, so they are indeed connected. In fact, in principle the speed of _indexing_ could also slow down consensus, with the way it's hooked up right now (though in practice I think that's unlikely to happen). I misunderstood the dependency order.

In effect, this means the pubsub system is the spinal column of the entire consensus engine.  That clearly wasn't the intent, but the idea that the event bus is the primary mechanism for synchronization isn't entirely irrational if we are willing to design around that. The way things are wired right now it would be a little tricky to decouple them, but I think it's doable.

That said, I'm becoming more and more convinced the event system needs to be given more weight in the design space, for all the reasons we've discussed here and elsewhere.